src/picongpu/include/particles/ionization/byField/ADK/ADK_Impl.hpp

```diff
@@ -0,0 +1,201 @@
+            typedef FieldB::ValueType ValueType_B;
+            /* global memory EM-field device databoxes */
+            PMACC_ALIGN(eBox, FieldE::DataBoxType);
+            PMACC_ALIGN(bBox, FieldB::DataBoxType);
+            /* shared memory EM-field device databoxes */
+            PMACC_ALIGN(cachedE, DataBox<SharedBox<ValueType_E, typename BlockArea::FullSuperCellSize,1> >);
+            PMACC_ALIGN(cachedB, DataBox<SharedBox<ValueType_B, typename BlockArea::FullSuperCellSize,0> >);
+
+        public:
+            /* host constructor initializing member : random number generator */
+            ADK_Impl(const uint32_t currentStep) : randomGen(currentStep)
+            {
+                DataConnector &dc = Environment<>::get().DataConnector();
+                /* initialize pointers on host-side E-(B-)field databoxes */
+                FieldE* fieldE = &(dc.getData<FieldE > (FieldE::getName(), true));
+                FieldB* fieldB = &(dc.getData<FieldB > (FieldB::getName(), true));
+                /* initialize device-side E-(B-)field databoxes */
+                eBox = fieldE->getDeviceDataBox();
+                bBox = fieldB->getDeviceDataBox();
+
+            }
+
+            /** Initialization function on device
+             *
+             * \brief Cache EM-fields on device
+             *         and initialize possible prerequisites for ionization, like e.g. random number generator.
+             *
+             * This function will be called inline on the device which must happen BEFORE threads diverge
+             * during loop execution. The reason for this is the `__syncthreads()` call which is necessary after
+             * initializing the E-/B-field shared boxes in shared memory.
+             */
+            DINLINE void init(const DataSpace<simDim>& blockCell, const int& linearThreadIdx, const DataSpace<simDim>& totalCellOffset)
+            {
+
+                /* caching of E and B fields */
+                cachedB = CachedBox::create < 0, ValueType_B > (BlockArea());
+                cachedE = CachedBox::create < 1, ValueType_E > (BlockArea());
+                /* wait for shared memory to be initialized */
+                __syncthreads();
+                /* instance of nvidia assignment operator */
+                nvidia::functors::Assign assign;
+                /* copy fields from global to shared */
+                PMACC_AUTO(fieldBBlock, bBox.shift(blockCell));
+                ThreadCollective<BlockArea> collective(linearThreadIdx);
+                collective(
+                          assign,
+                          cachedB,
+                          fieldBBlock
+                          );
+                /* copy fields from global to shared */
+                PMACC_AUTO(fieldEBlock, eBox.shift(blockCell));
+                collective(
+                          assign,
+                          cachedE,
+                          fieldEBlock
+                          );
+
```

Comment:
please call `__syncthreads();` heare to be sure that the shared memory operation before are finished.